File: src/main/java/fi/aalto/cs/apluscourses/utils/parser/S2FeedbackParser.java
package fi.aalto.cs.apluscourses.utils.parser;

import java.util.regex.Pattern;
import org.jetbrains.annotations.NotNull;
import org.jsoup.Jsoup;

public class S2FeedbackParser extends FeedbackParser {

  public static final String NAME = "S2";

  /**
   * Parses TestResults for S2 from an HTML string.
   */
  @Override
  @NotNull
  public FeedbackParser.TestResults parseTestResults(@NotNull String htmlString) {
    var body = Jsoup.parseBodyFragment(htmlString).body();
    var lastP = body.select("p").last();
    if (lastP == null) {
      return new FeedbackParser.TestResults(-1, -1);
    }
    var results = lastP.text();
    var pattern = Pattern.compile("success:\\s+(\\d+),\\s+failed:\\s+(\\d+)");
    var matcher = pattern.matcher(results);
    if (!matcher.find()) {
      return new FeedbackParser.TestResults(-1, -1);
    }
    return new FeedbackParser.TestResults(Integer.parseInt(matcher.group(1)), Integer.parseInt(matcher.group(2)));
  }
}

File: src/main/kotlin/fi/aalto/cs/apluscourses/utils/parser/S2FeedbackParser.kt
package fi.aalto.cs.apluscourses.utils.parser

import fi.aalto.cs.apluscourses.utils.parser.FeedbackParser.TestResults
import org.jsoup.Jsoup
import java.util.regex.Pattern

class S2FeedbackParser : FeedbackParser() {
    /**
     * Parses TestResults for S2 from an HTML string.
     */
    override fun parseTestResults(htmlString: String): TestResults {
        val body = Jsoup.parseBodyFragment(htmlString).body()
        val lastP = body.select("p").last()
        if (lastP == null) {
            return TestResults(-1, -1)
        }
        val results = lastP.text()
        val pattern = Pattern.compile("success:\\s+(\\d+),\\s+failed:\\s+(\\d+)")
        val matcher = pattern.matcher(results)
        if (!matcher.find()) {
            return TestResults(-1, -1)
        }
        return TestResults(matcher.group(1).toInt(), matcher.group(2).toInt())
    }

    companion object {
        const val NAME: String = "S2"
    }
}
